titus-common/src/main/java/com/netflix/titus/common/util/archaius2/Archaius2ObjectConfigurationResolver.java

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.archaius2;
+
+import java.io.Closeable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.Lists;
+import com.netflix.archaius.api.Config;
+import com.netflix.archaius.api.ConfigListener;
+import com.netflix.titus.common.util.PropertiesExt;
+import com.netflix.titus.common.util.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class Archaius2ObjectConfigurationResolver<OBJECT, CONFIG> implements ObjectConfigurationResolver<OBJECT, CONFIG>, ConfigListener, Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(Archaius2ObjectConfigurationResolver.class);
+
+    private final Config configuration;
+    private final Function<OBJECT, String> selectorFieldAccessor;
+    private final Class<CONFIG> configType;
+    private final CONFIG defaultConfig;
+
+    private volatile SortedMap<String, Pair<Pattern, CONFIG>> configMap = Collections.emptySortedMap();
+
+    Archaius2ObjectConfigurationResolver(Config configuration,
+                                         Function<OBJECT, String> selectorFieldAccessor,
+                                         Class<CONFIG> configType,
+                                         CONFIG defaultConfig) {
+        this.configuration = configuration;
+        this.selectorFieldAccessor = selectorFieldAccessor;
+        this.configType = configType;
+        this.defaultConfig = defaultConfig;
+
+        configuration.addListener(this);
+        doUpdate();
+    }
+
+    @Override
+    public void close() {
+        configuration.removeListener(this);
+    }
+
+    @Override
+    public CONFIG resolve(OBJECT object) {
+        if (object == null) {
+            return defaultConfig;
+        }
+
+        String selectorValue = selectorFieldAccessor.apply(object);
+        if (selectorValue == null) {
+            return defaultConfig;
+        }
+
+
+        for (Map.Entry<String, Pair<Pattern, CONFIG>> entry : configMap.entrySet()) {
+            Pattern pattern = entry.getValue().getLeft();
+            if (pattern.matcher(selectorValue).matches()) {
+                return entry.getValue().getRight();
+            }
+        }
+        return defaultConfig;
+    }
+
+    @Override
+    public void onConfigAdded(Config config) {
+        doUpdate();
+    }
+
+    @Override
+    public void onConfigRemoved(Config config) {
+        doUpdate();
+    }
+
+    @Override
+    public void onConfigUpdated(Config config) {
+        doUpdate();
+    }
+
+    @Override
+    public void onError(Throwable error, Config config) {
+    }
+
+    private void doUpdate() {
```

Comment:
The Archaius API does not defined i there, but my assumption is that the callbacks are serialized, and in the right order. Otherwise I would consider this a bug.